File: ios/Runner/FlutterVpnPlugin.swift
import Flutter
import Foundation
import NetworkExtension
import UIKit

/// Flutter plugin for iOS.
/// Implements VpnHostApiProtocol (Pigeon) and bridges to VpnManager.
/// EventChannels emit push events for status, stats, and logs.
class FlutterVpnPlugin: NSObject, FlutterPlugin, VpnHostApiProtocol {

    private var statusChannel: FlutterEventChannel?
    private var statsChannel:  FlutterEventChannel?
    private var logChannel:    FlutterEventChannel?

    private var statusSink: FlutterEventSink?
    private var statsSink:  FlutterEventSink?
    private var logSink:    FlutterEventSink?

    private var statusTimer: Timer?
    private var statsTimer:  Timer?

    // ── Registration ──────────────────────────────────────────────────────

    static func register(with registrar: FlutterPluginRegistrar) {
        let plugin = FlutterVpnPlugin()
        VpnHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: plugin)

        plugin.statusChannel = FlutterEventChannel(
            name: "com.example.fluttervpngo/vpn_status_events",
            binaryMessenger: registrar.messenger())
        plugin.statsChannel = FlutterEventChannel(
            name: "com.example.fluttervpngo/vpn_stats_events",
            binaryMessenger: registrar.messenger())
        plugin.logChannel = FlutterEventChannel(
            name: "com.example.fluttervpngo/vpn_log_events",
            binaryMessenger: registrar.messenger())

        plugin.statusChannel?.setStreamHandler(
            EventStreamHandler { sink in plugin.statusSink = sink }
                               cancel: { plugin.statusSink = nil })
        plugin.statsChannel?.setStreamHandler(
            EventStreamHandler { sink in plugin.statsTimer = plugin.startStatsTimer(sink: sink) }
                               cancel: { plugin.statsTimer?.invalidate(); plugin.statsSink = nil })
        plugin.logChannel?.setStreamHandler(
            EventStreamHandler { sink in plugin.logSink = sink }
                               cancel: { plugin.logSink = nil })

        // Observe NE status changes
        VpnManager.shared.statusChangeHandler = { [weak plugin] status in
            plugin?.pushStatus(status)
        }
    }

    // ── prepareVpn ────────────────────────────────────────────────────────

    func prepareVpn(completion: @escaping (Result<Bool, Error>) -> Void) {
        Task { @MainActor in
            do {
                let ok = try await VpnManager.shared.prepare()
                completion(.success(ok))
            } catch {
                completion(.failure(error))
            }
        }
    }

    // ── installOrUpdateProfile ────────────────────────────────────────────

    func installOrUpdateProfile(config: TunnelConfigDto, completion: @escaping (Result<Void, Error>) -> Void) {
        Task { @MainActor in
            do {
                // Build config JSON (authToken comes from Keychain in extension)
                let configDict: [String: Any] = [
                    "id": config.id, "name": config.name,
                    "serverHost": config.serverHost, "serverPort": config.serverPort,
                    "transportType": config.transportType,
                    "deviceId": config.deviceId, "mtu": config.mtu,
                    "dnsServers": config.dnsServers.compactMap { $0 },
                    "ipv4Address": config.ipv4Address, "ipv4SubnetMask": config.ipv4SubnetMask,
                    "allowedIPv4Routes": config.allowedIPv4Routes.compactMap { $0 },
                    "excludedIPv4Routes": config.excludedIPv4Routes.compactMap { $0 },
                    "keepAliveSeconds": config.keepAliveSeconds, "workers": config.workers,
                    "sni": config.sni ?? "",
                    "enableUdp": config.enableUdp, "enableTcp": config.enableTcp,
                    "autoReconnect": config.autoReconnect,
                    "maxReconnectAttempts": config.maxReconnectAttempts,
                ]
                let data = try JSONSerialization.data(withJSONObject: configDict)
                let json = String(data: data, encoding: .utf8) ?? "{}"

                // Save authToken to Keychain if present
                if let token = config.authToken, !token.isEmpty {
                    KeychainHelper.shared.save(key: "authToken_\(config.id)", value: token)
                }

                try await VpnManager.shared.installOrUpdateProfile(
                    configJSON: json,
                    serverAddress: config.serverHost,
                    name: config.name
                )
                completion(.success(()))
            } catch {
                completion(.failure(error))
            }
        }
    }

    // ── startTunnel ───────────────────────────────────────────────────────

    func startTunnel(completion: @escaping (Result<Void, Error>) -> Void) {
        Task { @MainActor in
            do {
                try VpnManager.shared.startTunnel()
                completion(.success(()))
            } catch {
                completion(.failure(error))
            }
        }
    }

    // ── stopTunnel ────────────────────────────────────────────────────────

    func stopTunnel(completion: @escaping (Result<Void, Error>) -> Void) {
        Task { @MainActor in
            VpnManager.shared.stopTunnel()
            completion(.success(()))
        }
    }

    // ── getStatus ─────────────────────────────────────────────────────────

    func getStatus(completion: @escaping (Result<TunnelStatusDto, Error>) -> Void) {
        Task { @MainActor in
            let status = VpnManager.shared.currentStatus
            let uptime = VpnManager.shared.uptimeSeconds.map { Int64($0) }

            // Read server/transport from shared container
            let configDict = readConfigDict()
            let host = configDict?["serverHost"] as? String
            let transport = configDict?["transportType"] as? String

            completion(.success(TunnelStatusDto(
                status: status.tunnelStatusString,
                lastError: nil,
                serverHost: host,
                transportType: transport,
                uptimeSeconds: uptime
            )))
        }
    }

    // ── getStats ──────────────────────────────────────────────────────────

    func getStats(completion: @escaping (Result<TrafficStatsDto, Error>) -> Void) {
        Task {
            if let data = try? VpnManager.shared.readSharedFile(VpnConstants.statsKey),
               let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                completion(.success(TrafficStatsDto(
                    bytesIn:      Int64(dict["bytesIn"] as? Int ?? 0),
                    bytesOut:     Int64(dict["bytesOut"] as? Int ?? 0),
                    packetsIn:    Int64(dict["packetsIn"] as? Int ?? 0),
                    packetsOut:   Int64(dict["packetsOut"] as? Int ?? 0),
                    activeStreams: Int64(dict["activeStreams"] as? Int ?? 0),
                    updatedAtMs:  Int64(Date().timeIntervalSince1970 * 1000)
                )))
            } else {
                completion(.success(TrafficStatsDto(
                    bytesIn: 0, bytesOut: 0, packetsIn: 0, packetsOut: 0,
                    activeStreams: 0,
                    updatedAtMs: Int64(Date().timeIntervalSince1970 * 1000))))
            }
        }
    }

    // ── getLogs ───────────────────────────────────────────────────────────

    func getLogs(completion: @escaping (Result<[LogEntryDto?], Error>) -> Void) {
        Task {
            guard let data = try? VpnManager.shared.readSharedFile(VpnConstants.logsKey) else {
                completion(.success([])); return
            }
            let lines = String(data: data, encoding: .utf8)?
                .split(separator: "\n", omittingEmptySubsequences: true) ?? []
            var entries: [LogEntryDto?] = []
            for line in lines {
                if let dict = try? JSONSerialization.jsonObject(with: Data(line.utf8)) as? [String: Any] {
                    entries.append(LogEntryDto(
                        id:          dict["id"] as? String ?? UUID().uuidString,
                        timestampMs: Int64(dict["timestampMs"] as? Int ?? 0),
                        level:       dict["level"] as? String ?? "info",
                        category:    dict["category"] as? String ?? "extension",
                        message:     dict["message"] as? String ?? "",
                        repeatCount: Int64(dict["repeatCount"] as? Int ?? 1)
                    ))
                }
            }
            completion(.success(entries))
        }
    }

    // ── clearLogs ─────────────────────────────────────────────────────────

    func clearLogs(completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            try? VpnManager.shared.writeSharedFile(VpnConstants.logsKey, data: Data())
            completion(.success(()))
        }
    }

    // ── getDiagnostics ────────────────────────────────────────────────────

    func getDiagnostics(completion: @escaping (Result<DiagnosticsDto, Error>) -> Void) {
        Task { @MainActor in
            let bundle = Bundle.main
            let version = bundle.infoDictionary?["CFBundleShortVersionString"] as? String ?? "unknown"
            let build   = bundle.infoDictionary?["CFBundleVersion"] as? String ?? "unknown"
            let bundleId = bundle.bundleIdentifier ?? "unknown"
            let device   = UIDevice.current
            let osVersion = "\(device.systemName) \(device.systemVersion)"
            let model     = device.model

            let vpnStatus  = VpnManager.shared.currentStatus
            let neStatus: String
            switch vpnStatus {
            case .invalid:       neStatus = "notInstalled"
            case .disconnected:  neStatus = "installed"
            case .connecting, .connected, .reasserting, .disconnecting: neStatus = "enabled"
            @unknown default:    neStatus = "unknown"
            }

            // Try to get Go client version from shared container
            let goVersion: String
            if let data = try? VpnManager.shared.readSharedFile("go_version.txt"),
               let v = String(data: data, encoding: .utf8) {
                goVersion = v.trimmingCharacters(in: .whitespacesAndNewlines)
            } else {
                goVersion = "unavailable (build extension first)"
            }

            completion(.success(DiagnosticsDto(
                platform:                "ios",
                appVersion:              version,
                buildNumber:             build,
                osVersion:               osVersion,
                deviceModel:             model,
                vpnPermissionGranted:    vpnStatus != .invalid,
                networkExtensionStatus:  neStatus,
                runnerBundleId:          bundleId,
                extensionBundleId:       VpnConstants.extensionBundleId,
                appGroupId:              VpnConstants.appGroupId,
                goClientVersion:         goVersion,
                keychainAccessGroup:     VpnConstants.keychainAccessGroup
            )))
        }
    }

    // ── sendProviderMessage ───────────────────────────────────────────────

    func sendProviderMessage(type: String, payload: [String: Any?], completion: @escaping (Result<Void, Error>) -> Void) {
        Task { @MainActor in
            do {
                var msg: [String: Any] = ["type": type]
                for (k, v) in payload { if let v = v { msg[k] = v } }
                let data = try JSONSerialization.data(withJSONObject: msg)
                _ = try await VpnManager.shared.sendProviderMessage(data)
                completion(.success(()))
            } catch {
                completion(.failure(error))
            }
        }
    }

    // ── Push helpers ──────────────────────────────────────────────────────

    @MainActor
    private func pushStatus(_ status: NEVPNStatus) {
        let configDict = readConfigDict()
        let map: [String: Any?] = [
            "status": status.tunnelStatusString,
            "lastError": nil,
            "serverHost": configDict?["serverHost"] as? String,
            "transportType": configDict?["transportType"] as? String,
            "uptimeSeconds": VpnManager.shared.uptimeSeconds,
        ]
        statusSink?(map)
    }

    private func startStatsTimer(sink: @escaping FlutterEventSink) -> Timer {
        statsSink = sink
        return Timer.scheduledTimer(withTimeInterval: 2.0, repeats: true) { [weak self] _ in
            guard let self = self else { return }
            Task {
                if let data = try? VpnManager.shared.readSharedFile(VpnConstants.statsKey),
                   let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                    let map: [String: Any] = [
                        "bytesIn":      dict["bytesIn"] ?? 0,
                        "bytesOut":     dict["bytesOut"] ?? 0,
                        "packetsIn":    dict["packetsIn"] ?? 0,
                        "packetsOut":   dict["packetsOut"] ?? 0,
                        "activeStreams": dict["activeStreams"] ?? 0,
                        "updatedAtMs":  Int(Date().timeIntervalSince1970 * 1000),
                    ]
                    DispatchQueue.main.async { self.statsSink?(map) }
                }
            }
        }
    }

    private func readConfigDict() -> [String: Any]? {
        guard let data = try? VpnManager.shared.readSharedFile(VpnConstants.configKey) else { return nil }
        return try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    }
}

// ── EventStreamHandler helper ─────────────────────────────────────────────

private class EventStreamHandler: NSObject, FlutterStreamHandler {
    let onListen: (FlutterEventSink?) -> Void
    let onCancelAction: () -> Void

    init(onListen: @escaping (FlutterEventSink?) -> Void, cancel: @escaping () -> Void) {
        self.onListen = onListen
        self.onCancelAction = cancel
    }

    func onListen(withArguments args: Any?, eventSink sink: @escaping FlutterEventSink) -> FlutterError? {
        onListen(sink); return nil
    }
    func onCancel(withArguments args: Any?) -> FlutterError? {
        onCancelAction(); return nil
    }
}

File: android/vpnplugin/src/main/kotlin/com/example/fluttervpngo/vpnplugin/FlutterVpnPlugin.kt
package com.example.fluttervpngo.vpnplugin

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.VpnService
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.PluginRegistry
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import org.json.JSONObject

/**
 * Flutter plugin implementing the Pigeon VpnHostApi.
 *
 * Android-specific notes:
 *  - Commands (prepareVpn, startTunnel, stopTunnel, etc.) delegate to AndroidVpnService.
 *  - Status / stats / logs are read from AndroidVpnService static state and
 *    GoProcessBridge log buffer.
 *  - App exclusions are read/written via AppExclusionsHelper (SharedPreferences).
 *  - EventChannels push status/stats/log events to Flutter reactively.
 */
class FlutterVpnPlugin :
    FlutterPlugin,
    ActivityAware,
    PluginRegistry.ActivityResultListener,
    VpnHostApi {

    companion object {
        private const val TAG = "FlutterVpnPlugin"
        private const val VPN_REQUEST_CODE = 0x0F00

        private const val STATUS_CHANNEL = "com.example.fluttervpngo/vpn_status_events"
        private const val STATS_CHANNEL  = "com.example.fluttervpngo/vpn_stats_events"
        private const val LOG_CHANNEL    = "com.example.fluttervpngo/vpn_log_events"
    }

    private var context: Context? = null
    private var activity: Activity? = null
    private var activityBinding: ActivityPluginBinding? = null
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    // EventChannel sinks
    private var statusSink: EventChannel.EventSink? = null
    private var statsSink:  EventChannel.EventSink? = null
    private var logSink:    EventChannel.EventSink? = null

    private var prepareCallback: ((Result<Boolean>) -> Unit)? = null
    private var pendingConfigJson: String? = null

    // ── FlutterPlugin ─────────────────────────────────────────────────────────

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        VpnHostApi.setUp(binding.binaryMessenger, this)

        EventChannel(binding.binaryMessenger, STATUS_CHANNEL).setStreamHandler(
            object : EventChannel.StreamHandler {
                override fun onListen(a: Any?, s: EventChannel.EventSink?) { statusSink = s }
                override fun onCancel(a: Any?) { statusSink = null }
            }
        )
        EventChannel(binding.binaryMessenger, STATS_CHANNEL).setStreamHandler(
            object : EventChannel.StreamHandler {
                override fun onListen(a: Any?, s: EventChannel.EventSink?) { statsSink = s }
                override fun onCancel(a: Any?) { statsSink = null }
            }
        )
        EventChannel(binding.binaryMessenger, LOG_CHANNEL).setStreamHandler(
            object : EventChannel.StreamHandler {
                override fun onListen(a: Any?, s: EventChannel.EventSink?) { logSink = s }
                override fun onCancel(a: Any?) { logSink = null }
            }
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        VpnHostApi.setUp(binding.binaryMessenger, null)
        context = null
    }

    // ── ActivityAware ─────────────────────────────────────────────────────────

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        activityBinding = binding
        binding.addActivityResultListener(this)
    }
    override fun onDetachedFromActivity() {
        activityBinding?.removeActivityResultListener(this)
        activity = null; activityBinding = null
    }
    override fun onDetachedFromActivityForConfigChanges() = onDetachedFromActivity()
    override fun onReattachedToActivityForConfigChanges(b: ActivityPluginBinding) = onAttachedToActivity(b)

    // ── ActivityResultListener ────────────────────────────────────────────────

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
        if (requestCode != VPN_REQUEST_CODE) return false
        prepareCallback?.let { cb ->
            prepareCallback = null
            cb(Result.success(resultCode == Activity.RESULT_OK))
        }
        return true
    }

    // ── VpnHostApi ────────────────────────────────────────────────────────────

    override fun prepareVpn(callback: (Result<Boolean>) -> Unit) {
        val ctx = context ?: return callback(Result.failure(Exception("No context")))
        val intent = VpnService.prepare(ctx)
        if (intent == null) {
            callback(Result.success(true))
        } else {
            val act = activity ?: return callback(Result.failure(Exception("No activity for VPN dialog")))
            prepareCallback = callback
            act.startActivityForResult(intent, VPN_REQUEST_CODE)
        }
    }

    override fun installOrUpdateProfile(config: TunnelConfigDto, callback: (Result<Unit>) -> Unit) {
        try {
            pendingConfigJson = configDtoToJson(config)
            callback(Result.success(Unit))
        } catch (e: Exception) {
            callback(Result.failure(e))
        }
    }

    override fun startTunnel(callback: (Result<Unit>) -> Unit) {
        val ctx = context ?: return callback(Result.failure(Exception("No context")))
        val cfg = pendingConfigJson ?: return callback(Result.failure(Exception("No config. Call installOrUpdateProfile first.")))
        try {
            val intent = Intent(ctx, AndroidVpnService::class.java).apply {
                action = AndroidVpnService.ACTION_START
                putExtra(AndroidVpnService.EXTRA_CONFIG, cfg)
            }
            ctx.startForegroundService(intent)
            callback(Result.success(Unit))
            pushStatus("connecting")
        } catch (e: Exception) {
            Log.e(TAG, "startTunnel: ${e.message}", e)
            callback(Result.failure(e))
        }
    }

    override fun stopTunnel(callback: (Result<Unit>) -> Unit) {
        val ctx = context ?: return callback(Result.failure(Exception("No context")))
        try {
            ctx.startService(Intent(ctx, AndroidVpnService::class.java).apply {
                action = AndroidVpnService.ACTION_STOP
            })
            callback(Result.success(Unit))
        } catch (e: Exception) {
            callback(Result.failure(e))
        }
    }

    override fun getStatus(callback: (Result<TunnelStatusDto>) -> Unit) {
        val status = AndroidVpnService.currentStatus
        val uptime = if (status == "connected" && AndroidVpnService.connectedAt > 0L)
            ((System.currentTimeMillis() - AndroidVpnService.connectedAt) / 1000).toLong()
        else null
        val cfg = pendingConfigJson
        val host      = cfg?.let { safeJson(it, "serverHost") }
        val transport = cfg?.let { safeJson(it, "transportType") }

        callback(Result.success(TunnelStatusDto(
            status        = status,
            lastError     = AndroidVpnService.lastError,
            serverHost    = host,
            transportType = transport,
            uptimeSeconds = uptime,
        )))
        // Push to EventChannel as well
        pushStatus(status, AndroidVpnService.lastError)
    }

    override fun getStats(callback: (Result<TrafficStatsDto>) -> Unit) {
        val bridge = AndroidVpnService.logBridgeInstance
        if (bridge != null) {
            callback(Result.success(TrafficStatsDto(
                bytesIn      = bridge.statsBytesIn.get(),
                bytesOut     = bridge.statsBytesOut.get(),
                packetsIn    = bridge.statsPktsIn.get(),
                packetsOut   = bridge.statsPktsOut.get(),
                activeStreams = bridge.activeWorkers.get().toLong(),
                updatedAtMs  = System.currentTimeMillis(),
            )))
        } else {
            callback(Result.success(TrafficStatsDto(0,0,0,0,0, System.currentTimeMillis())))
        }
    }

    override fun getLogs(callback: (Result<List<LogEntryDto?>>) -> Unit) {
        val entries = AndroidVpnService.logBridgeInstance?.drainLogs() ?: emptyList()
        callback(Result.success(entries.map { e ->
            LogEntryDto(e.id, e.timestampMs, e.level, e.category, e.message, e.repeatCount.toLong())
        }))
    }

    override fun clearLogs(callback: (Result<Unit>) -> Unit) {
        AndroidVpnService.logBridgeInstance?.clearLogs()
        callback(Result.success(Unit))
    }

    override fun getDiagnostics(callback: (Result<DiagnosticsDto>) -> Unit) {
        val ctx = context ?: return callback(Result.failure(Exception("No context")))
        try {
            val pm = ctx.packageManager
            val pkgInfo = pm.getPackageInfo(ctx.packageName, 0)
            val vpnGranted = VpnService.prepare(ctx) == null
            val build = android.os.Build.VERSION.RELEASE
            val sdk   = android.os.Build.VERSION.SDK_INT
            val model = "${android.os.Build.MANUFACTURER} ${android.os.Build.MODEL}"

            callback(Result.success(DiagnosticsDto(
                platform               = "android",
                appVersion             = pkgInfo.versionName ?: "unknown",
                buildNumber            = pkgInfo.longVersionCode.toString(),
                osVersion              = "Android $build (SDK $sdk)",
                deviceModel            = model,
                vpnPermissionGranted   = vpnGranted,
                networkExtensionStatus = "n/a",
                runnerBundleId         = ctx.packageName,
                extensionBundleId      = "${ctx.packageName}.PacketTunnel",
                appGroupId             = "n/a",
                goClientVersion        = "1.0.0-mock (process)",
                keychainAccessGroup    = "n/a",
            )))
        } catch (e: Exception) {
            callback(Result.failure(e))
        }
    }

    override fun sendProviderMessage(type: String, payload: Map<String, Any?>, callback: (Result<Unit>) -> Unit) {
        Log.d(TAG, "sendProviderMessage: $type / $payload")
        callback(Result.success(Unit))
    }

    // ── App exclusions API ────────────────────────────────────────────────────

    override fun getInstalledApps(callback: (Result<List<AppInfoDto?>>) -> Unit) {
        val ctx = context ?: return callback(Result.failure(Exception("No context")))
        scope.launch(Dispatchers.IO) {
            val apps = AppExclusionsHelper.getInstalledApps(ctx)
            val excluded = AppExclusionsHelper.loadExclusions(ctx)
            val isWhitelist = AppExclusionsHelper.loadIsWhitelist(ctx)
            scope.launch(Dispatchers.Main) {
                callback(Result.success(apps.map { a ->
                    AppInfoDto(
                        packageName = a.packageName,
                        label       = a.label,
                        iconBase64  = a.iconBase64,
                        isExcluded  = excluded.contains(a.packageName),
                        isWhitelist = isWhitelist,
                    )
                }))
            }
        }
    }

    override fun setExcludedApps(packages: List<String?>, isWhitelist: Boolean, callback: (Result<Unit>) -> Unit) {
        val ctx = context ?: return callback(Result.failure(Exception("No context")))
        val set = packages.filterNotNull().toSet()
        AppExclusionsHelper.saveExclusions(ctx, set)
        AppExclusionsHelper.saveIsWhitelist(ctx, isWhitelist)
        // Live-reload WireGuard tunnel if running
        scope.launch(Dispatchers.IO) {
            // Signal service to reload
            ctx.startService(Intent(ctx, AndroidVpnService::class.java).apply {
                action = "com.example.fluttervpngo.ACTION_RELOAD_WG"
            })
        }
        callback(Result.success(Unit))
    }

    // ── Push helpers ──────────────────────────────────────────────────────────

    private fun pushStatus(status: String, error: String? = null) {
        val cfg = pendingConfigJson
        val map = mapOf<String, Any?>(
            "status"        to status,
            "lastError"     to error,
            "serverHost"    to cfg?.let { safeJson(it, "serverHost") },
            "transportType" to cfg?.let { safeJson(it, "transportType") },
        )
        activity?.runOnUiThread { statusSink?.success(map) }
    }

    // ── Serialization ─────────────────────────────────────────────────────────

    private fun configDtoToJson(dto: TunnelConfigDto): String = JSONObject().apply {
        put("id",                   dto.id)
        put("name",                 dto.name)
        put("serverHost",           dto.serverHost)
        put("serverPort",           dto.serverPort)
        put("transportType",        dto.transportType)
        put("authToken",            dto.authToken ?: "")
        put("deviceId",             dto.deviceId)
        put("mtu",                  dto.mtu)
        put("dnsServers",           org.json.JSONArray(dto.dnsServers.filterNotNull()))
        put("ipv4Address",          dto.ipv4Address)
        put("ipv4SubnetMask",       dto.ipv4SubnetMask)
        put("allowedIPv4Routes",    org.json.JSONArray(dto.allowedIPv4Routes.filterNotNull()))
        put("excludedIPv4Routes",   org.json.JSONArray(dto.excludedIPv4Routes.filterNotNull()))
        put("keepAliveSeconds",     dto.keepAliveSeconds)
        put("workers",              dto.workers)
        put("sni",                  dto.sni ?: "")
        put("enableUdp",            dto.enableUdp)
        put("enableTcp",            dto.enableTcp)
        put("autoReconnect",        dto.autoReconnect)
        put("maxReconnectAttempts", dto.maxReconnectAttempts)
    }.toString()

    private fun safeJson(json: String, key: String): String? =
        runCatching { JSONObject(json).optString(key).ifEmpty { null } }.getOrNull()
}
